### PR TITLE
Updating keytar typings for v4.0.2

### PR DESCRIPTION
**Pull request by juturu**
Updating typing for keytar 4.0.2 

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/atom/node-keytar/pull/63

### CODE REVIEW

**Comment by juturu** [comment]
@shiftkey @miniak @binarymuse please take a look.

File: types/keytar/index.d.ts [hunk 1]
```diff
@@ -21,39 +21,25 @@ export declare function getPassword(service: string, account: string): string |
  * @param account The string account name.
  * @param password The string password.
  *
- * @returns true on success, false on failure.
+ * @returns Promise, which on success yields empty string or error string on failure.
  */
-export declare function addPassword(service: string, account: string, password: string): boolean;
+export declare function setPassword(service: string, account: string, password: string): Promise<string>;
```
Comment (shiftkey) [inline types/keytar/index.d.ts]: I believe `setPassword` is now a `Promise<void>`

File: types/keytar/index.d.ts [hunk 2]
```diff
@@ -21,39 +21,25 @@ export declare function getPassword(service: string, account: string): string |
...
- * followed by addPassword(service, account, password).
- *
- * @param service The string service name.
- * @param account The string account name.
- * @param password The string password.
- *
- * @returns true on success, false on failure.
- */
-export declare function replacePassword(service: string, account: string, password: string): boolean;
+export declare function deletePassword(service: string, account: string): Promise<string>;
```
Comment (shiftkey) [inline types/keytar/index.d.ts]: `deletePassword` now returns `Promise<boolean>` to indicate success

File: types/keytar/keytar-tests.ts [hunk 3]
```diff
@@ -1,13 +1,12 @@
 
 import keytar = require('keytar');
 
-let success: boolean = false;
+let success: Promise<string>;
```
Comment (shiftkey) [inline types/keytar/keytar-tests.ts]: This should be `Promise<boolean>`

File: types/keytar/index.d.ts [hunk 4]
```diff
@@ -10,9 +10,9 @@
  * @param service The string service name.
  * @param account The string account name.
  *
- * @returns the string password or null on failures.
+ * @returns Promise, which on success yields the password string or error string on failures.
```
Comment (shiftkey) [inline types/keytar/index.d.ts]: I think this message can be tidied up - the error message should be included in the promise when it fails, and isn't specialized to what keytar is doing here.

**Review by shiftkey — CHANGES_REQUESTED** [review]
On the right track, just some return types here I think are off.

cc @binarymuse as I haven't had a chance to bump keytar and test these out myself

**Comment by juturu** [comment]
@shiftkey thanks for the review, i have updated the return types.

**Comment by BinaryMuse** [comment]
Not super familiar with TypeScript, but assuming `Promise<string>` indicates that the resolve value might be null, this looks right to me. (Both `getPassword` and `findPassword` resolve with `null` if a password couldn't be found.)

**Comment by shiftkey** [comment]
@BinaryMuse TS won't let you create a null string, but this can be represented as `Promise<string | null>` if that's a result we want to propagate, rather than letting the promise fail.

**Comment by BinaryMuse** [comment]
@shiftkey Gotcha. In keytar, the Promise will only be rejected if there was an actual unhandle-able exception getting the promise (down in the native implementations, e.g. the macOS keychain couldn't be accessed or other such fatal errors). The password not existing isn't included in this.

**Comment by shiftkey** [comment]
@BinaryMuse cool, I'll have another read of the docs after this lands and ensure they reflect these new changes :metal:

**Comment by juturu** [comment]
@BinaryMuse unable to find/get password in key chain should also be treated as failure and promise needs to be rejected. isn't it?
I have updated the promise here for v4.0.2 to return `Promise<string | null>`

**Comment by BinaryMuse** [comment]
> unable to find/get password in key chain should also be treated as failure

@juturu I'm *personally* of the opinion that it should not be. In the context of "exceptions should be exceptional", a password not existing in a password management system isn't really an error — it's just an entry that doesn't yet exist (much like a dictionary with a missing key).

**Comment by juturu** [comment]
I think this change is ready to be merged. Who can help push this change in?

Thanks @BinaryMuse @shiftkey for help!

**Comment by shiftkey** [comment]
@juturu thanks for the work getting this updated, and thanks @BinaryMuse for the assist on the review.

**Comment by shiftkey** [comment]
:thumbsup: this is good to go

**Pull request merged**